File: 0733-flood-fill/0733-flood-fill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        if(image == null || image.length == 0){
            return image;
        }

        int original = image[sr][sc];
        helper(image,sr,sc,original,color);
        return image;
    }
    private void helper(int[][] grid,int row, int col,int original,int color){
        if(row < 0 || row >= grid.length || col < 0 
        || col >= grid[0].length || grid[row][col] != original || grid[row][col] == color){
            return;
        }
        
        
        grid[row][col] = color;
        helper(grid,row+1,col,original,color);
        helper(grid,row-1,col,original,color);
        helper(grid,row,col+1,original,color);
        helper(grid,row,col-1,original,color);
        
    }
}

File: 0733-flood-fill/0733-flood-fill.py
class Solution(object):
    def floodFill(self, image, sr, sc, color):
        """
        :type image: List[List[int]]
        :type sr: int
        :type sc: int
        :type color: int
        :rtype: List[List[int]]
        """
        if image is None or len(image) == 0:
            return image
        
        original = image[sr][sc]
        self.dfs(image,sr,sc,original,color)

        return image
    
    def dfs(self,grid,row,col,original,color):
        if row < 0 or row >= len(grid) or col < 0 or col >= len(grid[0]) or grid[row][col] != original or grid[row][col] == color:
            return
        
        grid[row][col] = color
        self.dfs(grid,row+1,col,original,color)
        self.dfs(grid,row-1,col,original,color)
        self.dfs(grid,row,col+1,original,color)
        self.dfs(grid,row,col-1,original,color)